File: Java/com/leetcode/binarySearchTreeIterator/BinarySearchTreeIterator.java
package com.leetcode.binarySearchTreeIterator;

import java.util.Stack;

import com.leetcode.dataStructures.TreeNode;

public class BinarySearchTreeIterator {

}

class BSTIterator {
	// Inorder traversal
    Stack<TreeNode> s = new Stack<TreeNode>();
    TreeNode pt = null;
    public BSTIterator(TreeNode root) {
        pt = root;
        while (pt != null) {
            s.push(pt);
            pt = pt.left;
        }
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        if (!s.empty() || pt != null) {
            return true;
        } else {
            return false;
        }
    }

    /** @return the next smallest number */
    public int next() {
        int value = -1;
        pt = s.pop();
        value = pt.val;
        pt = pt.right;
        while (pt != null) {
            s.push(pt);
            pt = pt.left;
        }
        return value;
    }
}

File: Python/binary_search_tree_iterator/binary_search_tree_iterator.py
# stack
class BSTIterator(object):
    def __init__(self, root):
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    def hasNext(self):
        return self.stack != []

    def next(self):
        self.curr = self.stack[-1]
        if not self.stack[-1].right:
            while self.stack and self.stack[-1].val <= self.curr.val:
                self.stack.pop()
        else:
            curr = self.stack[-1].right
            while curr:
                self.stack.append(curr)
                curr = curr.left

        return self.curr.val

# deconstruct steps
# setNext when necessary
class BSTIterator(object):
    def __init__(self, root):
        self.stack = []
        self.setNext(root)

    def hasNext(self):
        return self.stack != []

    def next(self):
        self.curr = self.stack.pop()
        self.setNext(self.curr.right)
        return self.curr.val

    def setNext(self, root):
        while root:
            self.stack.append(root)
            root = root.left
            